### DISCUSSION TITLE
Welcome to participle Discussions!

### DISCUSSION THREAD

**Discussion by alecthomas** [alecthomas/participle#126] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ThomasWidmann** [comment]
Hi, I am Thomas.

Have recently seen your parser implemented with go. Actually I have to cope with ldap based search filters and found your parser interesting. I am going to use this parser to interpret search filters but I think I haven't fully got how to  approach this.

## LDAP Filter
An ldap filter may look like this but could also be more complex using `&` (and)  and  `|`( or)

```
(&(objectClass=user)(objectCategory=person)(mail=a.b@cde.fg))
```

## Approach to Parse

I have created the struct below to parse the filter but I immediately recognized that this is not working because

1. _@Ident_ regular expression does create multiple tokens of mail address
2. `&` is seen as invalid character

 
**Struct**

```go
// Ldapfilters
type Ldapfilters struct {
	Filter []*Ldapfilter `@@*`
}

// Ldapfilter
type Ldapfilter struct {
	Filtertype string          `"(" @("&" | "|")`
	Properties []*LdapProperty `@@* ")"`
}

// LdapProperty
type LdapProperty struct {
	Key string `"(" @Ident "="`
	Val string `@Ident")"`
}
```

## Using Custom Lexer Definition
I tried to use a custom lever but I am not able to set it up properly. I am getting a compile error if I am getting a compile error when I do call participle.Build() or participle.MustBuild() - lever is not accepted. I think it is a type mismatch.

```go
// LdapProperty
type EmailProperty struct {
	Key string `"(" @Ident "="`
	Val string `@MyVal")"`
}

func TestEmailToken(t *testing.T) {

	basicLexer, err := stateful.NewSimple([]stateful.Rule{
		{"Comment", `(?i)rem[^\n]*`, nil},
		{"String", `"(\\"|[^"])*"`, nil},
		{"Number", `[-+]?(\d*\.)?\d+`, nil},
		{"Ident", `[a-zA-Z_.@]\w*`, nil},
		{"MyVal", `[a-zA-Z_][a-zA-Z0-9_.@]{0,}`, nil},
		{"Punct", `[-[!@#$%^&*()+_={}\|:;"'<,>.?/]|]`, nil},
		{"EOL", `[\n\r]+`, nil},
		{"whitespace", `[ \t]+`, nil},
	})
	p3 := "(mail=a.b@cde.fg)"
	parser, err := participle.Build(
		&EmailProperty{},
		participle.Lexer(basicLexer),
	)
	ep := &EmailProperty{}
	err = parser.ParseString(p3, ep)
```

Probably someone can enlighten me :)

**Reply by alecthomas** [reply]
You haven't said what the exact error/issue is?

**Reply by ThomasWidmann** [reply]
The problem was sitting in front of the computer. I lost track with trial and error at some point. I copied the Basic Lexer and simply adjusted the regular expression at the type "Ident" so that I can parse a mail address as a value. This works fine now. I think the parser is really good, what did you originally build it for?
